components/script/dom/htmlmediaelement.rs

```diff
@@ -230,86 +226,66 @@ impl HTMLMediaElement {
                         &window,
                     );
                 }
-            }
 
-            // previous ready state was ReadyState::HaveFutureData or more,
-            // and the new ready state is ReadyState::HaveCurrentData or less.
-            (ReadyState::HaveFutureData, ReadyState::HaveCurrentData) |
-            (ReadyState::HaveEnoughData, ReadyState::HaveCurrentData) |
-            (ReadyState::HaveFutureData, ReadyState::HaveMetadata) |
-            (ReadyState::HaveEnoughData, ReadyState::HaveMetadata) |
-            (ReadyState::HaveFutureData, ReadyState::HaveNothing) |
-            (ReadyState::HaveEnoughData, ReadyState::HaveNothing) => {
-                // TODO: timeupdate event logic + waiting
-            }
+                // Steps for the transition from HaveMetadata to HaveCurrentData
+                // or HaveFutureData also apply here, as per the next match
+                // expression.
+            },
+            (ReadyState::HaveFutureData, new) if new <= ReadyState::HaveCurrentData => {
+                // FIXME(nox): Queue a task to fire timeupdate and waiting
+                // events if the conditions call from the spec are met.
+
+                // No other steps are applicable in this case.
+                return;
+            },
 
             _ => (),
         }
 
-        // Step 1.
-        // If the new ready state is ReadyState::HaveFutureData or ReadyState::HaveEnoughData,
-        // then the relevant steps below must then be run also.
-        match (old_ready_state, ready_state) {
-            // Previous ready state was ReadyState::HaveCurrentData or less,
-            // and the new ready state is ReadyState::HaveFutureData.
-            (ReadyState::HaveCurrentData, ReadyState::HaveFutureData) |
-            (ReadyState::HaveMetadata, ReadyState::HaveFutureData) |
-            (ReadyState::HaveNothing, ReadyState::HaveFutureData) => {
-                task_source.queue_simple_event(
-                    self.upcast(),
-                    atom!("canplay"),
-                    &window,
-                );
+        if old_ready_state <= ReadyState::HaveCurrentData && ready_state >= ReadyState::HaveFutureData {
```

Comment:
This isn't idempotent, you weren't testing for `HaveEnoughData` before.

Reply:
This transition should always be done, and then the final one to `HaveEnoughData`.